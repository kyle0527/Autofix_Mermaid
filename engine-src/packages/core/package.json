{
  "name": "@diagrammender/core",
  "version": "0.3.0",
  "main": "dist/src/index.js",
  "types": "dist/src/index.d.ts",
  "dependencies": {
    "@diagrammender/types": "file:../types",
    "@diagrammender/parsers-python": "file:../parsers/python",
    "@diagrammender/analyzers": "file:../analyzers",
    "@diagrammender/emitters-mermaid": "file:../emitters/mermaid",
    "@diagrammender/fix-rules-mermaid-compat": "file:../fix-rules/mermaid-compat"
  },
  "scripts": { "build": "tsc -p tsconfig.json", "test": "echo no-tests" }
}
