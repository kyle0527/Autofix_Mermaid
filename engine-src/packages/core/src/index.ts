export * from './pipeline';
export * from './parsers';