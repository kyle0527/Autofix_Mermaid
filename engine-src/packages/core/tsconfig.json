{
  "extends": "../../tsconfig.base.json",
  "include": ["src"],
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "./dist",
    "baseUrl": ".",
    "paths": {
      "@diagrammender/types": ["../types/dist/index"],
      "@diagrammender/analyzers": ["../analyzers/dist/index"],
      "@diagrammender/emitters-mermaid": ["../emitters/mermaid/dist/index"],
      "@diagrammender/fix-rules-mermaid-compat": ["../fix-rules/mermaid-compat/dist/index"]
    }
  }
}
