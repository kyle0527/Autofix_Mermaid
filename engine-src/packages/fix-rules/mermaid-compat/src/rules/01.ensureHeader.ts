import { FixRule } from '../types';
export const EnsureHeader: FixRule = {
  name: 'EnsureHeader', priority: 10,
  run(code, ctx) {
    const trimmed = code.trim();
    // If no valid diagram header present, add one based on ctx.diagram
    if (!/^(flowchart|classDiagram|sequenceDiagram|stateDiagram|erDiagram)\b/.test(trimmed)) {
      const kind = ctx?.diagram || 'flowchart';
      let header = 'flowchart TD';
      if (kind === 'classDiagram') header = 'classDiagram';
      else if (kind === 'sequenceDiagram') header = 'sequenceDiagram';
      // default fallbacks remain flowchart TD for better layout determinism
      return { code: `${header}\n${code}`, notes: [`Added diagram header: ${header}`] };
    }
    return { code, notes: [] };
  }
};
