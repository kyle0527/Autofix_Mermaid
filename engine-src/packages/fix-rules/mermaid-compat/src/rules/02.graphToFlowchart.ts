import { FixRule } from '../types';
export const GraphToFlowchart: FixRule = {
  name: 'GraphToFlowchart', priority: 20,
  run(code, ctx) {
    if (/^graph\s+[A-Z]{2}/mi.test(code)) {
      const c = code.replace(/^graph\s+([A-Z]{2})/gmi, 'flowchart $1');
      return { code: c, notes: ['Converted graph to flowchart'] };
    }
    return { code, notes: [] };
  }
};
