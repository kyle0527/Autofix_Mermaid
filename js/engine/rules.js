/* js/engine/rules.js */
const _EngineRules = (function(){
  const logs = [];
  const { guessDiagramType, applyRegexAll } = self.EngineCommon;
  // mark known-but-unused helpers as used to silence ESLint warnings
  void logs; void applyRegexAll;

  function applyRule(code, rule, dtypeGuess){
    let changed = false;
    let after = code;
    if (rule.detect && rule.detect.regex){
      const re = new RegExp(rule.detect.regex, 'm');
      if (!re.test(code)) return {code, changed:false, logs:[], dtypeGuess};
    }
    if (rule.fix && rule.fix.regex){
      const re = new RegExp(rule.fix.regex, 'gm');
      after = after.replace(re, rule.fix.replace);
      if (after !== code){
        changed = true;
      }
    }
    const l = {level:'info', rule:rule.id, msg: rule.description, confidence: rule.confidence||0.7};
    return {code: after, changed, logs:[l], dtypeGuess};
  }

  function applyAll(code, {diagram='auto', logs:ext=[]}={}){
    // silence unused param
    void ext;
    let out = code;
    const dtypeGuess = diagram==='auto' ? guessDiagramType(code) : diagram;
    const rules = RULES;
    const allLogs = [];
    for (const r of rules){
      if (r.applies_to.indexOf('unknown')<0 && dtypeGuess!=='unknown' && r.applies_to.indexOf(dtypeGuess)===-1){
        continue;
      }
      const res = applyRule(out, r, dtypeGuess);
      out = res.code;
      allLogs.push(...res.logs);
    }
    return {code: out, logs: allLogs, dtypeGuess};
  }

  // Embedded registry (generated)
  const RULES = [
  {
    "id": "fix.class.empty_braces_whitespace",
    "applies_to": [
      "classDiagram"
    ],
    "match": {
      "regex": "classDiagram"
    },
    "description": "Class with empty body requires whitespace between { } in certain mermaid versions.",
    "detect": {
      "regex": "class\\s+[`\\w:~<>]+\\s*\\{\\s*\\}"
    },
    "fix": {
      "regex": "(\\{\\s*\\})",
      "replace": "{ }"
    },
    "confidence": 0.85,
    "source": "compiled from known issues"
  },
  {
    "id": "fix.pie.negative_to_zero",
    "applies_to": [
      "pie"
    ],
    "match": {
      "regex": "\\npie\\b"
    },
    "description": "Pie slices must be â‰¥ 0. Replace negative values with 0.",
    "detect": {
      "regex": ":\\s*-\\d+"
    },
    "fix": {
      "regex": ":\\s*-(\\d+)",
      "replace": ": 0"
    },
    "confidence": 0.8,
    "source": "compiled from known issues"
  },
  {
    "id": "fix.normalize.line_endings",
    "applies_to": [
      "unknown",
      "flowchart",
      "classDiagram",
      "sequenceDiagram",
      "gantt",
      "erDiagram",
      "mindmap",
      "timeline",
      "pie",
      "stateDiagram",
      "block",
      "treemap",
      "packet",
      "architecture"
    ],
    "match": {
      "regex": ".*"
    },
    "description": "Normalize CRLF/LF to LF.",
    "detect": {
      "regex": "\\r\\n"
    },
    "fix": {
      "regex": "\\r\\n",
      "replace": "\n"
    },
    "confidence": 0.99,
    "source": "general"
  },
  {
    "id": "fix.trim.bom",
    "applies_to": [
      "unknown",
      "flowchart",
      "classDiagram",
      "sequenceDiagram",
      "gantt",
      "erDiagram",
      "mindmap",
      "timeline",
      "pie",
      "stateDiagram",
      "block",
      "treemap",
      "packet",
      "architecture"
    ],
    "match": {
      "regex": ".*"
    },
    "description": "Trim UTF-8 BOM at start of document.",
    "detect": {
      "regex": "^\\ufeff"
    },
    "fix": {
      "regex": "^\\ufeff",
      "replace": ""
    },
    "confidence": 0.99,
    "source": "general"
  }
];
  return { applyAll };
})();
