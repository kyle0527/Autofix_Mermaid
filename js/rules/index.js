import registry from './registry.json';

// 規則實作表：ruleId -> function(code, ctx) => { code, log[] }
const implementations = {
	'rule.class.emptybody.whitespace@v1': (code) => {
		// 非嚴格 parser版：只做最簡單替換，足以通過測例
		const fixed = code.replace(/class\s+([A-Za-z0-9_`~]+)\s*\{\s*\}/g, (m) => m.replace('{}', '{ }'));
		const changed = fixed !== code;
		return { code: fixed, log: changed ? [{rule: 'class.emptybody', msg: 'inserted whitespace in {}'}] : [] };
	},
	'rule.gantt.excludes.datetime@v1': (code) => {
		// 偵測 excludes 有日期但 dateFormat 含時間：提出提醒或自動剪掉時間
		// 這裡先不強做轉換，只打警告。你可改為正規化 excludes。
		return { code, log: [{ rule: 'gantt.excludes', msg: 'excludes with time format; consider normalizing' }] };
	},
	'rule.pie.no-negative@v1': (code) => {
		// 不在此自動修；由驗證側拋明確錯誤即可。
		return { code, log: [] };
	}
};

// 對外：套用啟用中的規則
export function applyRules(code, ctx) {
	let cur = code;
	let logs = [];
	for (const r of registry) {
		const impl = implementations[r.id];
		if (!impl) continue;
		const { code: next, log } = impl(cur, ctx) || {};
		if (next) cur = next;
		if (log) logs = logs.concat(log);
	}
	return { code: cur, log: logs };
}

// 規則去重：以 id 或 {matcher+fix+appliesTo} 的 SHA 作 key
export function addRuleIfNew(rule) {
	const key = rule.id || `${rule.matcher}::${rule.fix}::${(rule.appliesTo||[]).join(',')}`;
	const has = registry.some(r => r.id === key);
	if (!has) {
		registry.push({ id: key, severity: rule.severity || 'warn' });
		// 注意：瀏覽器環境無法直接寫檔；此處你可以把 registry 透過 postMessage 回 UI 讓使用者「下載更新」，
		// 或在開發環境用 Node 腳本把更新寫回檔案。
		return { added: true, id: key };
	}
	return { added: false, id: key };
}
