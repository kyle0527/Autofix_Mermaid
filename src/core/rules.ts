export type DiagramKind =
  | 'flowchart'
  | 'classDiagram'
  | 'sequenceDiagram'
  | 'stateDiagram'
  | 'erDiagram'
  | 'gantt'
  | 'pie'
  | 'mindmap';

export interface RuleCtx {
  mmd: string;
  diag: DiagramKind;
}

export interface Rule {
  id: string;
  description: string;
  apply(ctx: RuleCtx): RuleCtx; // 純函式，不副作用
  severity?: 'info' | 'warn' | 'error';
  fixable?: boolean;
}

export const rule_legacy_to_flowchart: Rule = {
  id: 'legacy-syntax-upgrade',
  description: '將 graph 開頭升級為 flowchart',
  severity: 'info',
  fixable: true,
  apply: (ctx) => {
    if (ctx.diag !== 'flowchart') return ctx;
    const next = ctx.mmd.replace(/^\s*graph\b/gi, 'flowchart');
    return { ...ctx, mmd: next };
  }
};

export const rule_id_normalize: Rule = {
  id: 'id-normalize',
  description: '節點 id 僅允許 ASCII，其他以 _ 代替（示例版）',
  severity: 'warn',
  fixable: true,
  apply: (ctx) => {
    // 只處理節點 ID：匹配行首如  n0[  n1(  n2-->  等 ID token
    const lines = ctx.mmd.split(/\n/).map(line => {
      const m = line.match(/^\s*([A-Za-z0-9_\u0080-\uFFFF]+)(?=[\[(]|-->|===|:::|\s)/);
      if (!m) return line;
      const rawId = m[1];
      const norm = rawId.replace(/[^A-Za-z0-9_]/g, '_');
      if (norm === rawId) return line;
      return line.replace(rawId, norm);
    });
    return { ...ctx, mmd: lines.join('\n') };
  }
};

// TODO: 後續新增
// 下面為骨架規則：先放佔位，利於 UI 勾選與測試；後續補實作細節。

export const rule_flow_direction: Rule = {
  id: 'flow-direction',
  description: '若未指定方向，預設加上 flowchart LR',
  severity: 'info',
  fixable: true,
  apply: (ctx) => {
    if (ctx.diag !== 'flowchart') return ctx;
    if (/^flowchart\s+(LR|RL|TB|BT|TD)/m.test(ctx.mmd)) return ctx;
    if (!/^flowchart/m.test(ctx.mmd)) return ctx; // 不是 flowchart 開頭，略過
    const lines = ctx.mmd.split(/\n/);
    const idx = lines.findIndex(l => /^flowchart\b/.test(l));
    if (idx >= 0) lines[idx] = lines[idx] + ' LR';
    return { ...ctx, mmd: lines.join('\n') };
  }
};

export const rule_diagram_kind_infer: Rule = {
  id: 'diagram-kind-infer',
  description: '嘗試從內容推斷 diagram 種類（示例: 有 class 關鍵字 → classDiagram）',
  severity: 'info',
  fixable: false,
  apply: (ctx) => {
    // 目前僅示例：若包含 "classDiagram" 字樣則返回原樣；若檔頭為空可能未來插入
    // 真實實作會在更上游決定 diag，這裡暫不變更。
    return ctx;
  }
};

export const rule_dangling_node_prune: Rule = {
  id: 'dangling-node-prune',
  description: '移除無任何連線的節點（暫為占位，不修改）',
  severity: 'warn',
  fixable: true,
  apply: (ctx) => {
    // 未來可解析節點行，計算 in/out degree 再剔除，現階段不動
    return ctx;
  }
};

export const rule_label_truncate: Rule = {
  id: 'label-truncate',
  description: '過長標籤截斷（>40 chars）並加 …（占位）',
  severity: 'info',
  fixable: true,
  apply: (ctx) => {
    // 簡易示例：不實際截斷，後續再補
    return ctx;
  }
};

export function applyRules(
  ctx: RuleCtx,
  rules: Rule[]
): { ctx: RuleCtx; applied: string[] } {
  let cur = ctx;
  const applied: string[] = [];
  for (const r of rules) {
    const prev = cur.mmd;
    cur = r.apply(cur);
    if (cur.mmd !== prev) applied.push(r.id);
  }
  return { ctx: cur, applied };
}