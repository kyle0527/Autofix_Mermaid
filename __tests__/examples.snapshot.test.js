/// <reference types="jest" />
import { applyRules, rule_flow_direction, rule_legacy_to_flowchart, rule_id_normalize } from '../src/core/rules';
// 由於 NodeNext 模式，需要顯式副檔名；此處測試層可允許透過 ts-jest 處理，不再補 .js
const examples = [];
// 只加入前 4 個與最後 1 個，避免測試檔過大；其餘可後續再補充分檔維護
examples.push(`flowchart TD\n    n0([ Start: run_checks.py module ])\n    n1([ End / 結束 ])\n    n2["Import / 導入"]\n    n3["ImportFrom / 導入自"]\n    n4["sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))"]\n    n5["ImportFrom / 導入自"]\n    n6["def main... / 定義"]\n    n7{"if __name__ == '__main__' / 如果 __name__ == '__main__'"}\n    n8["main() / main()"]\n    n9["merge / 合併"]\n    n0 --> n2\n    n2 --> n3\n    n3 --> n4\n    n4 --> n5\n    n5 --> n6\n    n6 --> n7\n    n7 -->|True| n8\n    n7 -->|False| n9\n    n8 --> n9\n    n9 --> n1`);
examples.push(`flowchart TD\n    n0([" Start: main() / Start: main() "])\n    n1([" End / 結束 "])\n    n2["ap = argparse.ArgumentParser() / ap = argparse.ArgumentParser()"]\n    n3["ap.add_argument('--in', dest='infile', required=False, help='path to apis.json') / ap.add_argument('--in', dest='infile', required=False, help='path to apis.json')"]\n    n4["ap.add_argument('--checks', default='idor,oauth') / ap.add_argument('--checks', default='idor,oauth')"]\n    n5["ap.add_argument('--out', dest='outfile', default='out/findings.jsonl') / ap.add_argument('--out', dest='outfile', default='out/findings.jsonl')"]\n    n6["args = ap.parse_args() / args = ap.parse_args()"]\n    n7["checks = [c.strip() for c in args.checks.split(',') if c.strip()] / checks = [c.strip() for c in args.checks.split(',') if c.strip()]"]\n    n8["classes = [REGISTRY[c] for c in checks if c in REGISTRY] / classes = [REGISTRY[c] for c in checks if c in REGISTRY]"]\n    n9["apis = None / apis = None"]\n    n10{"if args.infile / 如果 args.infile"}\n    n11["try / 嘗試"]\n    n12["With / With"]\n    n13["except Exception / 例外"]\n    n14["apis = None / apis = None"]\n    n15["after try / try 後"]\n    n16["merge / 合併"]\n    n17["out_lines: List[str] = [] / out_lines: List[str] = []"]\n    n18{"for Cls in classes / for Cls in classes"}\n    n19["inst = Cls() / inst = Cls()"]\n    n20{"for finding in inst.run(apis=apis) / for finding in inst.run(apis=apis)"}\n    n21["out_lines.append(finding.to_json()) / out_lines.append(finding.to_json())"]\n    n22["after for / for 後"]\n    n23["after for / for 後"]\n    n24["Import / 導入"]\n    n25["os.makedirs(os.path.dirname(args.outfile), exist_ok=True) / os.makedirs(os.path.dirname(args.outfile), exist_ok=True)"]\n    n26["With / With"]\n    n27["print(f'wrote {len(out_lines)} findings to {args.outfile}') / print(f'wrote {len(out_lines)} findings to {args.outfile}')"]\n    n0 --> n2\n    n2 --> n3\n    n3 --> n4\n    n4 --> n5\n    n5 --> n6\n    n6 --> n7\n    n7 --> n8\n    n8 --> n9\n    n9 --> n10\n    n10 -->|True| n11\n    n10 -->|False| n16\n    n11 --> n12\n    n11 --> n13\n    n12 --> n15\n    n13 --> n14\n    n14 --> n15\n    n15 --> n16\n    n16 --> n17\n    n17 --> n18\n    n18 -->|True| n19\n    n18 -->|False| n23\n    n19 --> n20\n    n20 -->|True| n21\n    n20 -->|False| n22\n    n21 --> n20\n    n22 --> n18\n    n23 --> n24\n    n24 --> n25\n    n25 --> n26\n    n26 --> n27\n    n27 --> n1`);
examples.push(`flowchart TD\n    n0(["Start: scorecard.py (module) / Start: scorecard.py (module)"])\n    n1(["End / 結束"])\n    n2["Import / 導入"]\n    n3["WEIGHTS = {'discovery': 0.2, 'protocol': 0.25, 'verification': 0.15, 'ranking': 0.1, 'delivery': 0.1, 'ecosystem': 0.1, 'compliance': 0.05, 'performance': 0.05} / WEIGHTS = {'discovery': 0.2, 'protocol': 0.25, 'verification': 0.15, 'ranking': 0.1, 'delivery': 0.1, 'ecosystem': 0.1, 'compliance': 0.05, 'performance': 0.05}"]\n    n4["def score(...) / 定義"]\n    n5["def main(...) / 定義"]\n    n6{"if __name__ == '__main__' / 如果 __name__ == '__main__'"}\n    n7["main() / main()"]\n    n8["merge / 合併"]\n    n0 --> n2\n    n2 --> n3\n    n3 --> n4\n    n4 --> n5\n    n5 --> n6\n    n6 -->|True| n7\n    n6 -->|False| n8\n    n7 --> n8\n    n8 --> n1`);
examples.push(`flowchart TD\n    n0["Start: post_rank(ranked)"]\n    n1["End / 結束"]\n    n2["Tag findings created by :func:\`vector_probe\`."]\n    n3{"for item in ranked"}\n    n4{"if item.get('check') == 'demo/entrypoint'"}\n    n5["item.setdefault('tags', []).append('entrypoint-demo')"]\n    n6["merge"]\n    n7["after for"]\n    n8["return ranked"]\n    n0 --> n2\n    n2 --> n3\n    n3 -->|True| n4\n    n3 -->|False| n7\n    n4 -->|True| n5\n    n4 -->|False| n6\n    n5 --> n6\n    n6 --> n3\n    n7 --> n8\n    n8 --> n1`);
examples.push(`flowchart TD\n    n0(["Start: _extract_paths(text)"])\n    n1(["End / 結束"])\n    n2["paths: List[str] = []"]\n    n3{"for line in text.splitlines()"}\n    n4["stripped = line.strip()"]\n    n5{"if stripped.startswith('/') and stripped.endswith(':')"}\n    n6["paths.append(stripped[:-1])"]\n    n7["merge / 合併"]\n    n8["after for / for 後"]\n    n9["return paths / 返回 paths"]\n    n0 --> n2\n    n2 --> n3\n    n3 -->|True| n4\n    n3 -->|False| n8\n    n4 --> n5\n    n5 -->|True| n6\n    n5 -->|False| n7\n    n6 --> n7\n    n7 --> n3\n    n8 --> n9\n    n9 --> n1`);
describe('Provided example diagrams stability (subset)', () => {
    examples.forEach((ex, idx) => {
        test(`example subset #${idx + 1}`, () => {
            const input = { mmd: ex, diag: 'flowchart' };
            const { ctx, applied } = applyRules(input, [
                rule_legacy_to_flowchart,
                rule_flow_direction,
                rule_id_normalize
            ]);
            expect({ mmd: ctx.mmd, appliedCount: applied.length }).toMatchSnapshot();
        });
    });
});
