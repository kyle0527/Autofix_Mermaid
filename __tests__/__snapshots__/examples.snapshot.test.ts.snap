// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provided example diagrams stability (subset) example subset #1 1`] = `
{
  "appliedCount": 0,
  "mmd": "flowchart TD
    n0([ Start: run_checks.py module ])
    n1([ End / 結束 ])
    n2["Import / 導入"]
    n3["ImportFrom / 導入自"]
    n4["sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))"]
    n5["ImportFrom / 導入自"]
    n6["def main... / 定義"]
    n7{"if __name__ == '__main__' / 如果 __name__ == '__main__'"}
    n8["main() / main()"]
    n9["merge / 合併"]
    n0 --> n2
    n2 --> n3
    n3 --> n4
    n4 --> n5
    n5 --> n6
    n6 --> n7
    n7 -->|True| n8
    n7 -->|False| n9
    n8 --> n9
    n9 --> n1",
}
`;

exports[`Provided example diagrams stability (subset) example subset #2 1`] = `
{
  "appliedCount": 0,
  "mmd": "flowchart TD
    n0([" Start: main() / Start: main() "])
    n1([" End / 結束 "])
    n2["ap = argparse.ArgumentParser() / ap = argparse.ArgumentParser()"]
    n3["ap.add_argument('--in', dest='infile', required=False, help='path to apis.json') / ap.add_argument('--in', dest='infile', required=False, help='path to apis.json')"]
    n4["ap.add_argument('--checks', default='idor,oauth') / ap.add_argument('--checks', default='idor,oauth')"]
    n5["ap.add_argument('--out', dest='outfile', default='out/findings.jsonl') / ap.add_argument('--out', dest='outfile', default='out/findings.jsonl')"]
    n6["args = ap.parse_args() / args = ap.parse_args()"]
    n7["checks = [c.strip() for c in args.checks.split(',') if c.strip()] / checks = [c.strip() for c in args.checks.split(',') if c.strip()]"]
    n8["classes = [REGISTRY[c] for c in checks if c in REGISTRY] / classes = [REGISTRY[c] for c in checks if c in REGISTRY]"]
    n9["apis = None / apis = None"]
    n10{"if args.infile / 如果 args.infile"}
    n11["try / 嘗試"]
    n12["With / With"]
    n13["except Exception / 例外"]
    n14["apis = None / apis = None"]
    n15["after try / try 後"]
    n16["merge / 合併"]
    n17["out_lines: List[str] = [] / out_lines: List[str] = []"]
    n18{"for Cls in classes / for Cls in classes"}
    n19["inst = Cls() / inst = Cls()"]
    n20{"for finding in inst.run(apis=apis) / for finding in inst.run(apis=apis)"}
    n21["out_lines.append(finding.to_json()) / out_lines.append(finding.to_json())"]
    n22["after for / for 後"]
    n23["after for / for 後"]
    n24["Import / 導入"]
    n25["os.makedirs(os.path.dirname(args.outfile), exist_ok=True) / os.makedirs(os.path.dirname(args.outfile), exist_ok=True)"]
    n26["With / With"]
    n27["print(f'wrote {len(out_lines)} findings to {args.outfile}') / print(f'wrote {len(out_lines)} findings to {args.outfile}')"]
    n0 --> n2
    n2 --> n3
    n3 --> n4
    n4 --> n5
    n5 --> n6
    n6 --> n7
    n7 --> n8
    n8 --> n9
    n9 --> n10
    n10 -->|True| n11
    n10 -->|False| n16
    n11 --> n12
    n11 --> n13
    n12 --> n15
    n13 --> n14
    n14 --> n15
    n15 --> n16
    n16 --> n17
    n17 --> n18
    n18 -->|True| n19
    n18 -->|False| n23
    n19 --> n20
    n20 -->|True| n21
    n20 -->|False| n22
    n21 --> n20
    n22 --> n18
    n23 --> n24
    n24 --> n25
    n25 --> n26
    n26 --> n27
    n27 --> n1",
}
`;

exports[`Provided example diagrams stability (subset) example subset #3 1`] = `
{
  "appliedCount": 0,
  "mmd": "flowchart TD
    n0(["Start: scorecard.py (module) / Start: scorecard.py (module)"])
    n1(["End / 結束"])
    n2["Import / 導入"]
    n3["WEIGHTS = {'discovery': 0.2, 'protocol': 0.25, 'verification': 0.15, 'ranking': 0.1, 'delivery': 0.1, 'ecosystem': 0.1, 'compliance': 0.05, 'performance': 0.05} / WEIGHTS = {'discovery': 0.2, 'protocol': 0.25, 'verification': 0.15, 'ranking': 0.1, 'delivery': 0.1, 'ecosystem': 0.1, 'compliance': 0.05, 'performance': 0.05}"]
    n4["def score(...) / 定義"]
    n5["def main(...) / 定義"]
    n6{"if __name__ == '__main__' / 如果 __name__ == '__main__'"}
    n7["main() / main()"]
    n8["merge / 合併"]
    n0 --> n2
    n2 --> n3
    n3 --> n4
    n4 --> n5
    n5 --> n6
    n6 -->|True| n7
    n6 -->|False| n8
    n7 --> n8
    n8 --> n1",
}
`;

exports[`Provided example diagrams stability (subset) example subset #4 1`] = `
{
  "appliedCount": 0,
  "mmd": "flowchart TD
    n0["Start: post_rank(ranked)"]
    n1["End / 結束"]
    n2["Tag findings created by :func:\`vector_probe\`."]
    n3{"for item in ranked"}
    n4{"if item.get('check') == 'demo/entrypoint'"}
    n5["item.setdefault('tags', []).append('entrypoint-demo')"]
    n6["merge"]
    n7["after for"]
    n8["return ranked"]
    n0 --> n2
    n2 --> n3
    n3 -->|True| n4
    n3 -->|False| n7
    n4 -->|True| n5
    n4 -->|False| n6
    n5 --> n6
    n6 --> n3
    n7 --> n8
    n8 --> n1",
}
`;

exports[`Provided example diagrams stability (subset) example subset #5 1`] = `
{
  "appliedCount": 0,
  "mmd": "flowchart TD
    n0(["Start: _extract_paths(text)"])
    n1(["End / 結束"])
    n2["paths: List[str] = []"]
    n3{"for line in text.splitlines()"}
    n4["stripped = line.strip()"]
    n5{"if stripped.startswith('/') and stripped.endswith(':')"}
    n6["paths.append(stripped[:-1])"]
    n7["merge / 合併"]
    n8["after for / for 後"]
    n9["return paths / 返回 paths"]
    n0 --> n2
    n2 --> n3
    n3 -->|True| n4
    n3 -->|False| n8
    n4 --> n5
    n5 -->|True| n6
    n5 -->|False| n7
    n6 --> n7
    n7 --> n3
    n8 --> n9
    n9 --> n1",
}
`;
